selectrum.el

```diff
@@ -1128,14 +1128,34 @@ INHERIT-INPUT-METHOD, see `completing-read-multiple'."
   "Complete in-buffer text using a list of candidates.
 Can be used as `completion-in-region-function'. For START, END,
 COLLECTION, and PREDICATE, see `completion-in-region'."
-  (let ((cands (nconc
-                (completion-all-completions
-                 (buffer-substring-no-properties start end)
-                 collection
-                 predicate
-                 (- end start))
-                nil))
-        (result nil))
+  (let* ((cands (nconc
+                 (completion-all-completions
+                  (buffer-substring-no-properties start end)
+                  collection
+                  predicate
+                  (- end start))
+                 nil))
+         (result nil)
+         (annotation-func (plist-get completion-extra-properties
+                                     :annotation-function))
+         (docsig-func (plist-get completion-extra-properties
+                                 :company-docsig))
+         (selectrum-preprocess-candidates-function
+          (lambda (cands)
+            (selectrum--map-destructive
+             (lambda (cand)
+               (propertize cand
+                           'selectrum-candidate-display-prefix
+                           (when annotation-func
+                             (propertize
+                              (format "%-12s" (funcall annotation-func cand))
```

Comment:
Is the 12-char width from the original Emacs implementation or did you just pick a reasonable-seeming number?

Reply:
It's not in Emacs. I am only familiar with 2 capf backends: the one offered by elisp-mode, which only adds a ` <f>` for callable symbols (and I think only when they appear together with non-callable symbols); and the one created by myself, which adds those `(function)`, `(variable)` things. So I pick 12 mainly to fit my own need. I know you are familiar with lsp-mode, right? I suggest you test this on some language servers and see if it's reasonable.

Reply:
Can we put the annotation to the right of the candidate instead of to the left? Then, if any annotations are displayed, we can quickly compute the maximum length of any candidate, and put all the annotations at the appropriate column.

Reply:
I don' quite understand. Are you saying to put the annotation in the suffix? And, to get the max length of the candidates, and do proper left padding in the suffix? If we do that, a very long candidate can make every candidate become multiline. Though that wouldn't happen often, but using the prefix would never have this problem. You be the judge ;)

Reply:
Well, okay, what about putting the annotation in the suffix directly? It would not be table-aligned, but it seems more elegant. And that does not have the long-candidate problem that you point out.

Reply:
Done. And, in case you ask why there isn't a space between the candidate and the suffix: capf backends actually does this for us. The default capf interface displays annotation in the suffix, and doesn't insert a space in between, so backends will prefix the annotation using a space as a workaround.

Reply:
Sounds good, thanks for explaining!